_posts/2019-01-15-recursion.md

```diff
@@ -0,0 +1,313 @@
+
+```elixir
+defmodule MyList
+  def delete_all(list, el) do
+    # coming soon!
+  end
+end
+```
+
+However, we need access to a new, empty list that we'll populate with the elements of the original list we're _not_ deleting. So, we'll define a version of `delete_all` that takes in _three_ arguments: the original list, the element who occurrences we want to delete, and the new empty list.
+
+`MyList.delete_all/2` will call the `MyList.delete_all/3` function. This saves the user from having to call `delete_all` with a third argument of an empty list and allows us to provide a nice tidy API.
+
+```elixir
+defmodule MyList
+  def delete_all(list, el) do
+    delete_all(list, el, [])
+  end
+
+  def delete_all([head | list], el, new_list) do
+  end
+end
+```
+
+The `MyList.delete_all/3` function's first job is to determine whether or not the first element in the current list, the `head` of the list, is the same value as the element we want to remove.
+
+If so, we *won't* add it to our new list. Instead, we'll call `MyList.delete_all/3` again with the remainder of the current list, the `tail`, and pass in our `new_list` unchanged. We can accomplish this with a guard clause:
+
+```elixir
+def delete_all([head | tail], el, new_list) when head === el do
+  delete_all(tail, el, new_list)
+end
+```
+
+If the head of the current list is *not* equal to the value we want to remove, however, we *do* want to add it to our `new_list` before moving on.
+
+We'll define another `delete_all/3` function, this time without a guard clause, to meet this condition:
+
+```elixir
+def delete_all([head | tail], el, new_list) do
+  delete_all(tail, el, [head | new_list])
+end
+```
+
+We add the current `head` to our new list like this:
+
+```elixir
+[ head | new_list ]
+```
+
+and we call `delete_all/3` again, passing it the remainder of the list (`tail`), the element to delete and the updated `new_list`.
+
+When should we stop recursing? In other words, what is the base case that will cause us to stop calling `delete_all/3`? When we've recursed over all of the elements in the original list, such that the `tail` is empty, we'll _stop_ calling `delete_all/3` and instead return the new list. Let's define one final `delete_all/3` function to match this condition:
+
+```elixir
+def delete_all([], el, new_list) do
+  new_list
+end
+```
+
+The only problem with this approach is that it builds and returns a new list in which all of the elements we kept from the original list are populated in reverse order. This is because by building out our new list like this:
+
+```elixir
+[ head | new_list ]
+```
+
+We are adding the element we want to keep to the _front_ of our new list, instead of the end.
+
+We can fix this by using `Enum.reverse` on the `new_list` once we've reached our empty list base case:
+
+```elixir
+def delete_all([], el, new_list) do
+  new_list
```

Comment:
Keeping with the community styleguide can we avoid pipelines for single functions?

```suggestion
  Enum.reverse(new_list)
```